### ISSUE TITLE
BlocBuilder doesn't need to explicitly provide the State type again

### ISSUE THREAD

**Issue by bigworld12** [felangel/bloc#560]
i am not sure why we need to pass the state type again in BlocBuilder.
currently we would write
```dart
BlocBuilder<BlocA,StateA>(
builder: (BuildContext context, StateA state) => ...
)
```
but why can't we remove the StateA, and instead write

```dart
BlocBuilder<BlocA>(
builder: (BuildContext context, StateA state) => ...
)
```
since BlocA already defines the state type.
and checking the defination here
https://github.com/felangel/bloc/blob/805cf865fe0972e7f1bd24f08598d624066434fb/packages/flutter_bloc/lib/src/bloc_builder.dart#L18

why can't we just write it as 
```dart
class BlocBuilder<B extends Bloc<dynamic, S>> extends BlocBuilderBase<B, S>
```

i am not sure if this is a limitation in dart, but it's annoying that i have to pass the state type again for every BlocBuilder

**Comment by felangel** [comment]
Hi @bigworld12 👋 
Thanks for opening an issue!

Unfortunately, I don't think this is possible.

```dart
class BlocBuilder<B extends Bloc<dynamic, S>> extends BlocBuilderBase<B, S>
```
won't compile because you need to define `S` (unless I'm missing something).

I totally agree that it would be ideal not to have to pass `State` and that dart should be able to infer the type.

**Comment by tenhobi** [comment]
Maybe DartLangSpec can help, page 61, chapter 14 Generics, https://dart.dev/guides/language/specifications/DartLangSpec-v2.2.pdf.

**Comment by bigworld12** [comment]
yeah i think this is a language limitation, c# also has the same problem
valid :
```csharp
class BlocBuilder<TBloc,TState> where TBloc : Bloc<object,TState>
```

invalid:
```csharp
class BlocBuilder<TBloc> where TBloc : Bloc<object,TState>
```

**Comment by felangel** [comment]
@tenhobi yeah exactly. I think Dart just has no way to know what the generic type S is unless you define it unfortunately.

**Comment by mateusfccp** [comment]
Well, here goes another Dart flaw to my list.
